File: src/main/java/com/freya02/botcommands/api/utils/ButtonContent.java
package com.freya02.botcommands.api.utils;

import net.dv8tion.jda.api.entities.emoji.Emoji;
import net.dv8tion.jda.api.interactions.components.buttons.Button;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Represents the visual content of a {@link Button}, this contains either an {@link Emoji} or a {@link String}
 */
public record ButtonContent(String text, Emoji emoji) {
	/**
	 * Constructs a {@link ButtonContent} from a {@link String}
	 *
	 * @param text The {@link String} to put in the {@link Button}
	 * @return The {@link ButtonContent} with the string
	 */
	public static ButtonContent withString(@NotNull String text) {
		return new ButtonContent(text, null);
	}

	/**
	 * Constructs a {@link ButtonContent} from an {@link Emoji}
	 *
	 * @param emoji The {@link Emoji} to put in the {@link Button}
	 * @return The {@link ButtonContent} with the emoji
	 */
	public static ButtonContent withEmoji(@NotNull Emoji emoji) {
		return new ButtonContent(null, emoji);
	}

	/**
	 * Constructs a {@link ButtonContent} from a {@link String} and an {@link Emoji}
	 *
	 * @param emoji The {@link Emoji} to put in the {@link Button}
	 * @return The {@link ButtonContent} with the text and emoji
	 */
	public static ButtonContent withEmoji(@NotNull String text, @NotNull Emoji emoji) {
		return new ButtonContent(text, emoji);
	}

	/**
	 * Constructs a {@link ButtonContent} from an unicode emoji
	 *
	 * @param unicode The unicode emoji
	 * @return The {@link ButtonContent} with the unicode emoji
	 */
	public static ButtonContent withEmoji(@NotNull String unicode) {
		return new ButtonContent(null, Emoji.fromUnicode(unicode));
	}

	/**
	 * Constructs a {@link ButtonContent} from a {@link String} and an unicode emoji
	 *
	 * @param unicode The unicode emoji
	 * @return The {@link ButtonContent} with the text and unicode emoji
	 */
	public static ButtonContent withEmoji(@NotNull String text, @NotNull String unicode) {
		return new ButtonContent(text, Emoji.fromUnicode(unicode));
	}

	/**
	 * Constructs a {@link ButtonContent} from a shortcode emoji, such as {@code :joy:}
	 *
	 * @param shortcode The shortcode emoji
	 * @return The {@link ButtonContent} with the shortcode emoji
	 */
	public static ButtonContent withShortcode(@NotNull String shortcode) {
		return new ButtonContent(null, EmojiUtils.resolveJDAEmoji(shortcode));
	}

	/**
	 * Constructs a {@link ButtonContent} from a {@link String} and a shortcode emoji, such as {@code :joy:}
	 *
	 * @param shortcode The shortcode emoji
	 * @return The {@link ButtonContent} with the text and shortcode emoji
	 */
	public static ButtonContent withShortcode(@NotNull String text, @NotNull String shortcode) {
		return new ButtonContent(text, EmojiUtils.resolveJDAEmoji(shortcode));
	}

	@Nullable
	public String text() {
		return text;
	}

	@Nullable
	public Emoji emoji() {
		return emoji;
	}
}

File: src/main/kotlin/io/github/freya022/botcommands/api/components/utils/ButtonContent.kt
package io.github.freya022.botcommands.api.components.utils

import io.github.freya022.botcommands.api.components.Button
import io.github.freya022.botcommands.api.utils.EmojiUtils
import net.dv8tion.jda.api.entities.emoji.Emoji
import net.dv8tion.jda.api.interactions.components.buttons.ButtonStyle
import javax.annotation.CheckReturnValue

/**
 * Represents the visual content of a [Button], this contains at least an [Emoji] or a [String]
 */
data class ButtonContent(val style: ButtonStyle, val label: String?, val emoji: Emoji?) {
    init {
        require(label != null || emoji != null) { "A label or an emoji needs to be set" }

        if (label != null) {
            require(label.isNotEmpty()) {
                "The label cannot be empty"
            }
        }
    }

    /**
     * Creates a new button content with the provided emoji alias / emoji unicode.
     *
     * ### Example
     * ```kt
     * // Emoji alias
     * withEmoji(":smiley:")
     * // Unicode emoji
     * withEmoji("😃")
     *
     * // Animated custom emoji
     * withEmoji("<a:dance:123456789123456789>")
     * // Not animated custom emoji
     * withEmoji("<:dog:123456789123456789>")
     *
     * // Unicode emoji, escape codes
     * withEmoji("&#92;uD83D&#92;uDE03")
     * // Codepoint notation
     * withEmoji("U+1F602")
     * ```
     */
    @CheckReturnValue
    fun withEmoji(aliasOrUnicode: String?): ButtonContent {
        val newEmoji = aliasOrUnicode?.let {
            EmojiUtils.resolveJDAEmojiOrNull(it) ?: Emoji.fromFormatted(it)
        }

        return ButtonContent(style, label, newEmoji)
    }

    companion object {
        /**
         * Constructs a [ButtonContent] with a label.
         */
        @JvmStatic
        fun fromLabel(style: ButtonStyle, label: String): ButtonContent {
            return ButtonContent(style, label, null)
        }

        /**
         * Constructs a [ButtonContent] with an [Emoji].
         */
        @JvmStatic
        fun fromEmoji(style: ButtonStyle, emoji: Emoji): ButtonContent {
            return ButtonContent(style, null, emoji)
        }

        /**
         * Constructs a [ButtonContent] with a label and an [Emoji].
         */
        @JvmStatic
        fun fromEmoji(style: ButtonStyle, label: String, emoji: Emoji): ButtonContent {
            return ButtonContent(style, label, emoji)
        }

        /**
         * Constructs a [ButtonContent] with an unicode emoji,
         * see [Emoji.fromUnicode] for accepted formats.
         */
        @JvmStatic
        fun fromUnicode(style: ButtonStyle, unicode: String): ButtonContent {
            return ButtonContent(style, null, Emoji.fromUnicode(unicode))
        }

        /**
         * Constructs a [ButtonContent] with a label and an unicode emoji,
         * see [Emoji.fromUnicode] for accepted formats.
         */
        @JvmStatic
        fun fromUnicode(style: ButtonStyle, label: String, unicode: String): ButtonContent {
            return ButtonContent(style, label, Emoji.fromUnicode(unicode))
        }

        /**
         * Constructs a [ButtonContent] from a shortcode emoji, such as `:joy:`.
         */
        @JvmStatic
        fun fromShortcode(style: ButtonStyle, shortcode: String): ButtonContent {
            return ButtonContent(style, null, EmojiUtils.resolveJDAEmoji(shortcode))
        }

        /**
         * Constructs a [ButtonContent] from a [String] and a shortcode emoji, such as `:joy:`.
         */
        @JvmStatic
        fun fromShortcode(style: ButtonStyle, text: String, shortcode: String): ButtonContent {
            return ButtonContent(style, text, EmojiUtils.resolveJDAEmoji(shortcode))
        }
    }
}
